File: apps/patter-voice-sdk/libraries/python/getpatter/llm/hermes.py
"""Hermes agent-runtime LLM for Patter pipeline mode.

Thin preset over :class:`getpatter.llm.openai_compatible.OpenAICompatibleLLMProvider`
that defaults the base URL, model, env-key name, and timeout for the Hermes
agent runtime. Patter is the voice shell (carrier + STT + turn-taking + TTS);
Hermes is the brain on the line — each turn is one
``POST {base_url}/chat/completions`` against the local Hermes gateway.

Hermes is stateless and keys continuity off HEADERS, not the OpenAI ``user``
field. Patter sends ``X-Hermes-Session-Id: patter-call-<call_id>`` on every
turn so one phone call maps to one Hermes session / transcript (on by default).
For long-term memory scoping, set ``session_key`` to emit a static
``X-Hermes-Session-Key`` header (off by default). The OpenAI ``user`` field is
still sent (``patter-call-<call_id>``) as a harmless upstream-log correlation
id, but it is not what drives the session.
"""

from __future__ import annotations

import os
from typing import Callable, ClassVar

from getpatter.llm.openai_compatible import OpenAICompatibleLLMProvider
from getpatter.models import SessionContext

__all__ = ["LLM"]

# Hermes gateway default (loopback; operator-co-located deployment).
_BASE_URL = "http://127.0.0.1:8642/v1"
_DEFAULT_MODEL = "hermes-agent"

# Hermes is stateless — continuity is carried in headers.
_SESSION_USER_PREFIX = "patter-call-"
_SESSION_ID_HEADER = "X-Hermes-Session-Id"
_SESSION_ID_PREFIX = "patter-call-"
_SESSION_KEY_HEADER = "X-Hermes-Session-Key"


class LLM(OpenAICompatibleLLMProvider):
    """Hermes agent-runtime LLM provider.

    Example::

        from getpatter.llm import hermes

        llm = hermes.LLM()                         # all env-defaulted
        llm = hermes.LLM(model="hermes-7b")        # explicit model override
        llm = hermes.LLM(api_key="...", base_url="http://host:8642/v1")

    Defaults:

    * ``base_url`` → ``http://127.0.0.1:8642/v1``
    * ``model`` → ``API_SERVER_MODEL_NAME`` env if set, else ``"hermes-agent"``
    * ``api_key`` → ``api_key`` arg or ``API_SERVER_KEY`` env (may be absent for
      a keyless local Hermes)
    * ``timeout`` → ``120.0`` s (runtimes run tools / memory / skills before
      replying, so turns can take 30-90 s)
    * per-call continuity → ``X-Hermes-Session-Id: patter-call-<call_id>``
      (always sent with a call id — the primary mechanism)
    * long-term memory → ``X-Hermes-Session-Key: <session_key>`` (only sent
      when ``session_key`` / ``session_key_from`` / ``session_key_factory`` is
      configured)

    Args:
        session_key: Optional STATIC long-term memory scope. When set, every
            turn emits ``X-Hermes-Session-Key: <session_key>`` so Hermes
            namespaces persistent memory across calls. Credential-grade — never
            logged. ``None`` (default) means the header is not sent.
        session_key_from: Convenience selector for a built-in per-call key
            derivation. Set to ``"caller_hash"`` to derive the session key per
            call as ``f"patter-caller-{ctx.caller_hash}"`` (a stable,
            non-reversible hash of the caller — never the raw number), enabling
            per-caller cross-call memory. ``None`` (default) uses the static
            ``session_key`` path. Ignored when ``session_key_factory`` is given
            explicitly.
        session_key_factory: Custom callable deriving the
            ``X-Hermes-Session-Key`` value per call from a
            :class:`getpatter.models.SessionContext`. Takes precedence over both
            ``session_key`` and ``session_key_from``. A falsy return omits the
            header for that call. Credential-grade — never logged.
    """

    provider_key: ClassVar[str] = "hermes"

    def __init__(
        self,
        api_key: str | None = None,
        *,
        base_url: str = _BASE_URL,
        model: str | None = None,
        timeout: float = 120.0,
        session_key: str | None = None,
        session_key_from: str | None = None,
        session_key_factory: Callable[[SessionContext], str | None] | None = None,
        **kwargs,
    ) -> None:
        resolved_model = model or os.environ.get(
            "API_SERVER_MODEL_NAME", _DEFAULT_MODEL
        )
        # ``session_key_from`` / ``session_key_factory`` resolution (the
        # "caller_hash" selector, factory precedence, validation) lives in the
        # generic OpenAICompatibleLLMProvider — this preset just forwards.
        super().__init__(
            api_key=api_key,
            base_url=base_url,
            model=resolved_model,
            api_key_env="API_SERVER_KEY",
            timeout=timeout,
            session_user_prefix=_SESSION_USER_PREFIX,
            session_id_header=_SESSION_ID_HEADER,
            session_id_prefix=_SESSION_ID_PREFIX,
            session_key_header=_SESSION_KEY_HEADER,
            session_key=session_key,
            session_key_from=session_key_from,
            session_key_factory=session_key_factory,
            **kwargs,
        )

File: apps/patter-voice-sdk/libraries/typescript/src/llm/hermes.ts
/**
 * Hermes agent-runtime LLM preset for Patter's pipeline mode.
 *
 * Thin preset over {@link OpenAICompatibleLLMProvider}: defaults the base URL,
 * model, env-key name, timeout, and session-continuity prefix for the Hermes
 * agent runtime so a user just writes ``phone.agent({ llm: new hermes.LLM() })``.
 *
 * Hermes runs tools / memory / skills internally before replying, so a single
 * conversation turn can take 30-90 s — hence the 120 s default timeout. Hermes
 * is stateless and keys continuity off HEADERS, not the OpenAI ``user`` field:
 * the preset sends ``X-Hermes-Session-Id: patter-call-<callId>`` on every turn
 * for per-call session / transcript continuity (on by default), and optionally
 * ``X-Hermes-Session-Key: <sessionKey>`` for long-term memory scoping when you
 * pass ``sessionKey``. (It also still emits ``user=patter-call-<callId>`` for
 * upstream-log correlation, but that is not what drives the session.)
 */
import type { SessionContext } from '../types';
import {
  OpenAICompatibleLLMProvider,
  type OpenAICompatibleLLMOptions,
} from './openai-compatible';

/** Default Hermes agent-runtime base URL (loopback, operator-controlled). */
const BASE_URL = 'http://127.0.0.1:8642/v1';
/** Fallback model when neither ``model`` nor ``API_SERVER_MODEL_NAME`` is set. */
const DEFAULT_MODEL = 'hermes-agent';
/** Env var Hermes reads its bearer from. */
const API_KEY_ENV = 'API_SERVER_KEY';
/** Env var Hermes reads its model id from. */
const MODEL_ENV = 'API_SERVER_MODEL_NAME';
/** Per-call ``user`` prefix (upstream-log correlation; not the session driver). */
const SESSION_USER_PREFIX = 'patter-call-';
/** Header carrying the per-call session id (the primary continuity mechanism). */
const SESSION_ID_HEADER = 'X-Hermes-Session-Id';
/** Prefix for the session-id header value → ``X-Hermes-Session-Id: patter-call-<callId>``. */
const SESSION_ID_PREFIX = 'patter-call-';
/** Static header scoping long-term memory (sent only when ``sessionKey`` is set). */
const SESSION_KEY_HEADER = 'X-Hermes-Session-Key';
/** Default timeout (seconds): runtimes run tools before replying. */
const DEFAULT_TIMEOUT_S = 120;

/** Constructor options for the Hermes ``LLM`` preset. */
export interface HermesLLMOptions {
  /** Bearer token. Falls back to ``API_SERVER_KEY`` env var when omitted. */
  apiKey?: string;
  /** Override the Hermes base URL (rarely needed). */
  baseUrl?: string;
  /** Model id. Falls back to ``API_SERVER_MODEL_NAME`` env, then ``"hermes-agent"``. */
  model?: string;
  /** Per-request timeout in seconds. Default ``120``. */
  timeout?: number;
  /**
   * Static long-term memory scope. When set, emits ``X-Hermes-Session-Key`` so
   * Hermes scopes durable memory to this value across calls. ``undefined``
   * (default) means the header is not sent. Credential-grade — never logged.
   */
  sessionKey?: string;
  /**
   * Convenience selector for a built-in per-call key derivation. Set to
   * ``'caller_hash'`` to derive the session key per call as
   * ``` `patter-caller-${ctx.callerHash}` ``` (a stable, non-reversible hash of
   * the caller — never the raw number), enabling per-caller cross-call memory.
   * ``undefined`` (default) uses the static ``sessionKey`` path. Ignored when
   * ``sessionKeyFactory`` is given explicitly. Mirrors Python
   * ``session_key_from``.
   */
  sessionKeyFrom?: 'caller_hash';
  /**
   * Custom callback deriving the ``X-Hermes-Session-Key`` value per call from a
   * {@link SessionContext}. Takes precedence over both ``sessionKey`` and
   * ``sessionKeyFrom``. A falsy return omits the header for that call.
   * Credential-grade — never logged. Mirrors Python ``session_key_factory``.
   */
  sessionKeyFactory?: (ctx: SessionContext) => string | undefined;
  /** Extra headers merged after the SDK ``User-Agent``. */
  extraHeaders?: Record<string, string>;
  /** Sampling temperature [0, 2]. */
  temperature?: number;
  /** Max tokens in the assistant response (sent as ``max_completion_tokens``). */
  maxTokens?: number;
  /** OpenAI-style ``response_format`` for JSON mode / structured outputs. */
  responseFormat?: Record<string, unknown>;
  /** Whether to allow parallel tool calls. */
  parallelToolCalls?: boolean;
  /** ``"auto" | "none" | "required"`` or a specific tool object. */
  toolChoice?: string | Record<string, unknown>;
  /** Sampling seed for reproducible outputs. */
  seed?: number;
  /** Nucleus sampling cutoff in [0, 1]. */
  topP?: number;
  /** Penalty in [-2, 2] applied to repeated tokens. */
  frequencyPenalty?: number;
  /** Penalty in [-2, 2] applied to seen tokens. */
  presencePenalty?: number;
  /** Stop sequence(s). */
  stop?: string | string[];
}

/**
 * Hermes agent-runtime LLM provider (OpenAI-compatible, streaming).
 *
 * @example
 * ```ts
 * import * as hermes from "getpatter/llm/hermes";
 * const llm = new hermes.LLM();                       // env-defaulted, keyless OK
 * const llm = new hermes.LLM({ apiKey: "...", model: "hermes-7b" });
 * ```
 */
export class LLM extends OpenAICompatibleLLMProvider {
  static readonly providerKey = 'hermes';

  constructor(opts: HermesLLMOptions = {}) {
    const model = opts.model ?? process.env[MODEL_ENV] ?? DEFAULT_MODEL;
    // ``sessionKeyFrom`` / ``sessionKeyFactory`` resolution (the
    // 'caller_hash' selector, factory precedence, validation) lives in the
    // generic OpenAICompatibleLLMProvider — this preset just forwards.
    const options: OpenAICompatibleLLMOptions = {
      apiKey: opts.apiKey,
      apiKeyEnv: API_KEY_ENV,
      baseUrl: opts.baseUrl ?? BASE_URL,
      model,
      timeout: opts.timeout ?? DEFAULT_TIMEOUT_S,
      sessionUserPrefix: SESSION_USER_PREFIX,
      sessionIdHeader: SESSION_ID_HEADER,
      sessionIdPrefix: SESSION_ID_PREFIX,
      sessionKeyHeader: SESSION_KEY_HEADER,
      sessionKey: opts.sessionKey,
      sessionKeyFrom: opts.sessionKeyFrom,
      sessionKeyFactory: opts.sessionKeyFactory,
      extraHeaders: opts.extraHeaders,
      temperature: opts.temperature,
      maxTokens: opts.maxTokens,
      responseFormat: opts.responseFormat,
      parallelToolCalls: opts.parallelToolCalls,
      toolChoice: opts.toolChoice,
      seed: opts.seed,
      topP: opts.topP,
      frequencyPenalty: opts.frequencyPenalty,
      presencePenalty: opts.presencePenalty,
      stop: opts.stop,
    };
    super(options);
  }
}
